diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -940,26 +946,46 @@ int DifferentialFunctionComparator::cmpBasicBlocks(
             // it will be used in an inverse condition. Hence, we skip it here
             // and mark that it may be inverse-matching the condition that
             // has been originally mapped to the operand of the not operation.
-            if (isLogicalNot(&*InstL) || isLogicalNot(&*InstR)) {
+            if (config.Patterns.InverseConditions
+                && (isLogicalNot(&*InstL) || isLogicalNot(&*InstR))) {
+                // Note: This does not handle cases with a lot of `not`:
+                // %1 = cmp true, false | %1 = cmp true, false
+                //                      | %2 = not %1 (%2 = xor %1, true)
+                //                      | %3 = not %2 (%3 = xor %2, true)
+                //                      | %4 = not %3 (%4 = xor %3, true)
+                // br %1, l %T, l %F    | br %4, l %F, l %T
                 sn_mapL.erase(&*InstL);
                 sn_mapR.erase(&*InstR);
 
                 std::pair<const Value *, const Value *> matchingPair;
+                std::pair<const Value *, const Value *> oldMatchingPair;
```

Comment:
Nice catch! I completely missed this one.

I wonder if we could come up with better names.
Suggestion:
```suggestion
                std::pair<const Value *, const Value *> condPair;
                std::pair<const Value *, const Value *> origCondPair;
```
Maybe even a comment explaining the purpose of each of these would be useful.

Reply:
Thanks, I added the comment and used your suggestion, but I used `prevCondPair` instead of `origCondPair`, because if there are 2 or more `not` instructions the old pair will not be the original condition but it would look like something like this `{not ..., nullptr}` because it is not currently able to find the original instruction. 

I am thinking of fixing it but when I tried to compile some programs with negation (`!`), after compilation there were none `not` (`xor`) instructions. Also currently it would not be able to handle if `not` instructions would be on both sides. But I am not sure if it makes sense to fix this.

Reply:
Yeah, LLVM does a good job in removing successive negations. You could try to use `-O0` to see if it helps but we still apply some transformations in SimpLL which may remove those.

Anyways, since this is so hard to produce in the IR, it practically shouldn't occur in real code and therefore I don't think it's worth fixing.

Reply:
I did not manage to get LLVM IR containing `not` (xor) instruction even with `-O0`, so I will leave it like this.